org-roam-doctor.el

```diff
@@ -0,0 +1,199 @@
+(declare-function org-roam--get-roam-buffers "org-roam")
+(declare-function org-roam--list-all-files "org-roam")
+(declare-function org-roam--org-roam-file-p "org-roam")
+
+(cl-defstruct (org-roam-doctor-checker (:copier nil))
+  (name 'missing-checker-name)
+  (description "")
+  (actions nil))
+
+(defconst org-roam-doctor--checkers
+  (list
+   (make-org-roam-doctor-checker
+    :name 'org-roam-doctor-broken-links
+    :description "Fix broken links."
+    :actions '(("d" . ("Unlink" . org-roam-doctor--remove-link))
+               ("r" . ("Replace link" . org-roam-doctor--replace-link))
+               ("R" . ("Replace link (keep label)" . org-roam-doctor--replace-link-keep-label))))))
+
+(defun org-roam-doctor-broken-links (ast)
+  "Checker for detecting broken links.
+AST is the org-element parse tree."
+  (org-element-map ast 'link
+    (lambda (l)
+      (when (equal "file" (org-element-property :type l))
+        (let ((file (org-element-property :path l)))
+          (and (not (file-remote-p file))
+               (not (file-exists-p file))
+               (list (org-element-property :begin l)
+                     (format (if (org-element-lineage l '(link))
+                                 "Link to non-existent image file \"%s\"\
+ in link description"
+                               "Link to non-existent local file \"%s\"")
+                             file))))))))
+
+(defun org-roam-doctor--check (buffer checkers)
+  "Check BUFFER for errors.
+CHECKERS is the list of checkers used."
+  (with-current-buffer buffer
+    (save-excursion
+      (goto-char (point-min))
+      (let* ((ast (org-element-parse-buffer))
+             (errors (sort (cl-mapcan
+                            (lambda (c)
+                              (mapcar
+                               (lambda (report)
+                                 (list (set-marker (make-marker) (car report))
+                                       (nth 1 report) c))
+                               (save-excursion
+                                 (funcall
+                                  (org-roam-doctor-checker-name c)
+                                  ast))))
+                            checkers)
+                           #'car-less-than-car)))
+        (dolist (e errors)
+          (pcase-let ((`(,m ,msg ,checker) e))
+            (switch-to-buffer buffer)
+            (goto-char m)
+            (org-reveal)
+            (org-roam-doctor--resolve msg checker)
+            (set-marker m nil)))
+        errors))))
+
+;;; Actions
+(defun org-roam-doctor--recursive-edit ()
+  "Launch into a recursive edit."
+  (message "When you're done editing press C-M-c to continue.")
+  (recursive-edit))
+
+(defun org-roam-doctor--replace-link ()
+  "Replace the current link with a new link."
+  (unless (org-in-regexp org-link-bracket-re 1)
+    (user-error "No link at point"))
+  (save-excursion
+    (delete-region (match-beginning 0) (match-end 0))
```

Comment:
I think this should be done in a temp-buffer.  Otherwise, if you `C-g` in a middle of the `org-roam-insert` meant to fix it, the link is removed from the current buffer.

A better option might be to catch errors/interrupts with a `condition-case` which would revert the buffer to its pre-doctor state.

Reply:
`org-roam-insert` doesn't really work in a temp-buffer though.

Reply:
Then the second option I've mentioned might be best.

Reply:
done... seems a bit inefficient though.